Contact requests blocked

MrPetovan opened the pull request:
Fixes #6981 

The big change in this PR is `Model\Contact::addRelationship` returning a value, either `true` for relationship establishment success, `null` for relationship establishment pending, and `false` for relationship establishment failure.

This allowed me to move AP-specific code out of this method and properly add a contact block provision to it.

Side fixes/improvements:
- SQL error in admin summary
- JsonLD exception handling
- API call not implemented logging

annando commented:
I guess you have to return "null" on error to not accidentally trigger a rejection. But I don't really like moving the AP code out of this function. Compare it with functions like "createFromProbe" or "terminateFriendship". These are the "black boxes" that do the whole protocol specific contact management for us. They can also be called - for example - via the API or possibly some other places. I think that this relocating of code makes debugging more complicated.

nupplaphil commented on src/Module/Admin/Summary.php:
```diff
@@ -26,7 +26,7 @@ public static function content()
 		// are there MyISAM tables in the DB? If so, trigger a warning message
 		$warningtext = [];
-		if (DBA::count('`information_schema`.`tables`', ['engine' => 'myisam', 'table_schema' => DBA::databaseName()])) {
+		if (DBA::count('information_schema`.`tables', ['engine' => 'myisam', 'table_schema' => DBA::databaseName()])) {
```
This one doesn't seem right. Or is this a hack to count on system tables? If so, maybe there is a better solution directly in `DBA` possible?

nupplaphil replied:
Like a optional flag in `count(..., $schema = null)` and if not null, use it

annando replied:
In fact I just had a look at the code - it works that way but is ugly.

annando replied:
I suggest that I apply some changes to the database class for this.

annando replied:
I suggest doing it that way:
```
DBA::count(['information_schema' => 'tables'], ['engine' => 'myisam', 'table_schema' => DBA::databaseName()])
```

nupplaphil replied:
much better of course! :-)

MrPetovan replied:
Moved to #7174

MrPetovan commented:
> I guess you have to return "null" on error to not accidentally trigger a rejection.

Why? What can justify acknowledging a contact request when the contact ID is missing? If the public contact doesn't exist?

> I think that this relocating of code makes debugging more complicated.

Why so? On the contrary, having protocol-specific stuff in the model makes methods bigger with multiples code paths, increasing complexity. Don't take my word for it, I removed 10 protocol-specific lines from `Contact::addRelationship`, including 2 conditions and multiple DB requests to retrieve the contact protocol, and I replaced it with 3 lines with 1 condition and no extra DB requests. The original version wasn't even handling the `ActivityPub\Transmitter::sendContactReject` case.

Yes, I understand the convenience of seeing what every protocol is doing on contact actions in a single method, but it isn't sustainable and is harder to maintain since there isn't a clear separation of concern.

I'll take care of `Contact::terminateFriendship` as well as it is in the same situation. `Contact::createFromProbe` has other issues (like the `$interactive` parameter flag) that are outside of the scope of this issue.

annando commented:
BTW: The not sending of ```sendContactReject``` was intentionally. Sending a contact reject is optional in the protocol description, since possibly a user don't want the requesting person to know that this user rejected the contact request instead of ignoring it.

The problem with splitting these functions into several parts has got the problem that you never know if there possibly will come some new protocol that needs some very different treatment that we cannot handle that way.

I don't have that much paradigm when coding, but I really prefer the concept to have functions that are doing the whole stuff. This especially is much better when it comes to tasks that are possibly be called from different places like in the future by the API.

MrPetovan commented:
There are two ways of handling this:
- Either you have a massive function that handles every possible case. Each new case inflates the function length until it is too long/complex to be easily parseable by humans.
- Or you identify atomic behaviors, export them in separate functions, and only call them when you need. I believe that no matter what protocol comes along, the underlying Friendica way of adding a relationship between a user and a contact won’t change. This is an atomic, protocol-independent behavior that deserves to have its own function. Now each protocol can call it (or not) depending on the protocol-specific behavior.

annando commented on src/Model/Contact.php:
```diff
@@ -2120,17 +2120,33 @@ public static function updateSslPolicy(array $contact, $new_policy)
+		}
+
+		// Contact is blocked on node-level or user-level
+		if (!empty($pub_contact['blocked']) || !empty($contact['blocked'])) {
```
BTW: We do have ```isBlocked``` and ```isBlockedByUser``` - we should use that instead.

MrPetovan replied:
Fair enough.

MrPetovan replied:
Done. You were 100% right, I forgot about the server block and the `user-contact` table. Thanks for the pointer.

annando commented on src/Protocol/ActivityPub/Processor.php:
```diff
@@ -541,7 +541,14 @@ public static function followUser($activity)
-		Contact::addRelationship($owner, $contact, $item, '', false, $note);
+		$result = Contact::addRelationship($owner, $contact, $item, false, $note);
+		if ($result === false) {
+			ActivityPub\Transmitter::sendContactReject($item['author-link'], $item['author-id'], $owner['uid']);
```
I don't like sending a contact reject in case of a blocked account. The other person doesn't need to know if we just ignore the person or have blocked the person.

MrPetovan replied:
I think they do, because there is a difference between ignore and block. Do we have a way to ignore a contact without following them/them following us?

MrPetovan replied:
I removed the follow reject response until we sort this out.

tobiasd replied:
We can just hit the "ignore" button on the introduction they send us and they will never get a response (as long as you don't dive into the list of ignored introductions)

MrPetovan replied:
In this case (blocked contact or blocked node) the contact request won't be saved as an introduction at all. The question is if the remote contact should receive a follow rejection response or not in this case.

tobiasd commented:
> The question is if the remote contact should receive a follow
> rejection response or not in this case.

With diaspora* it is the case that an discarded introduction is send over and over again, hence there is only the ignore button.

What will happen if the blocke users request is just dropped? Will the other end send requests again and again?

Will it stop if the rejection was send?

Does the other account know it was blocked?

If not, I'd say just drop the request into /dev/null who I block or ignore should not be share I think.

MrPetovan commented:
Ok then this PR is ready in this case. I already removed the contentious follow rejection response.
